pkg/util/pki/parse.go

```diff
@@ -140,3 +140,161 @@ func DecodeX509CertificateRequestBytes(csrBytes []byte) (*x509.CertificateReques
 
 	return csr, nil
 }
+
+type PEMBundle struct {
+	CAPEM    []byte
+	ChainPEM []byte
+}
+
+type chainNode struct {
+	cert *x509.Certificate
+
+	issuer *chainNode
+}
+
+// ParseCertificateChainPEM will decode the bundle before passing to
+// ParseCertificateChain
+func ParseCertificateChainPEM(pembundle []byte) (*PEMBundle, error) {
+	certs, err := DecodeX509CertificateChainBytes(pembundle)
+	if err != nil {
+		return nil, err
+	}
+	return ParseCertificateChain(certs)
+}
+
+// ParseCertificateChain will return an x509 PEM encoded certificate chain and
+// CA. The CA may not be a true root, but the highest intermediate certificate.
+// Removes duplicate entries. The returned CA may be empty if a single
+// certificate was passed. Chain contains leaf certificate first. Removes
+// comments and white space. An error is returned if the passed bundle is not a
+// valid flat tree chain, the bundle is malformed, or the chain is broken.
+func ParseCertificateChain(certs []*x509.Certificate) (*PEMBundle, error) {
+	// De-duplicate certificates
+	for i := 0; i < len(certs)-1; i++ {
+		for j := 1; j < len(certs); j++ {
+			if i == j {
+				continue
+			}
+			if certs[i].Equal(certs[j]) {
+				certs = append(certs[:i], certs[i+1:]...)
+			}
+		}
+	}
+
+	var chains []*chainNode
+	// Build the nodes to the linked list
+	for i := range certs {
+		chains = append(chains, &chainNode{cert: certs[i]})
+	}
+
+	for {
+		// We have built the full chain
```

Comment:
suggestion: make it clear that this is the termination condition

```suggestion
		// Stop as soon as the chain is fully built.
```